OpenRA.Mods.Common/Traits/World/PathFinder.cs

```diff
@@ -125,95 +85,189 @@ public List<CPos> FindPath(PathSearch search)
 		{
 			using (new PerfSample("Pathfinder"))
 			{
-				using (search)
-				{
-					List<CPos> path = null;
+				var key = "FindPath" + search.Self.ActorID + "" + search.From.X + "" + search.From.Y + "" + search.Heuristic(search.From);
+				var cachedPath = cacheStorage.Retrieve<List<CPos>>(key);
 
-					while (!search.Queue.Empty)
-					{
-						var p = search.Expand(world);
-						if (search.Heuristic(p) == 0)
-						{
-							path = MakePath(search.CellInfo, p);
-							break;
-						}
-					}
+				if (cachedPath != null)
+					return cachedPath;
 
-					var dbg = world.WorldActor.TraitOrDefault<PathfinderDebugOverlay>();
-					if (dbg != null)
-						dbg.AddLayer(search.Considered.Select(p => new Pair<CPos, int>(p, search.CellInfo[p].MinCost)), search.MaxCost, search.Owner);
+				var pb = pathFinder.FindPath(search);
 
-					if (path != null)
-						return path;
-				}
+				cacheStorage.Store(key, pb, DateTime.Now.AddMilliseconds(MaxPathAge * 40), TimeSpan.Zero);
 
-				// no path exists
-				return EmptyPath;
+				return pb;
 			}
 		}
 
-		static List<CPos> MakePath(CellLayer<CellInfo> cellInfo, CPos destination)
+		public List<CPos> FindBidiPath(PathSearch fromSrc, PathSearch fromDest)
 		{
-			var ret = new List<CPos>();
-			var pathNode = destination;
+			using (new PerfSample("Pathfinder"))
+			{
+				var key = "FindBidiPath" + fromSrc.Self.ActorID + "" + fromDest.Self.ActorID + "" + fromSrc.From.X + "" + fromSrc.From.Y + "" +
+					fromSrc.Heuristic(fromSrc.From) + "" + fromDest.From.X + "" + fromDest.From.Y + "" + fromDest.Heuristic(fromSrc.From);
+				var cachedPath = cacheStorage.Retrieve<List<CPos>>(key);
 
-			while (cellInfo[pathNode].Path != pathNode)
+				if (cachedPath != null)
+					return cachedPath;
+
+				var pb = pathFinder.FindBidiPath(fromSrc, fromDest);
+
+				cacheStorage.Store(key, pb, DateTime.Now.AddMilliseconds(MaxPathAge * 40), TimeSpan.Zero);
```

Comment:
Involves some combination of reading the docs, the sources, and conducting
experiments.

On Tue, Jan 20, 2015 at 11:38 AM, Rydra <notifications@github.com> wrote:

> In OpenRA.Mods.Common/Traits/World/PathFinder.cs
> <https://github.com/OpenRA/OpenRA/pull/7358#discussion-diff-23191774>:
>
> >
> > -			while (cellInfo[pathNode].Path != pathNode)
> > +				if (cachedPath != null)
> > +					return cachedPath;
> > +
> > +				var pb = pathFinder.FindBidiPath(fromSrc, fromDest);
> > +
> > +				cacheStorage.Store(key, pb, DateTime.Now.AddMilliseconds(MaxPathAge * 40), TimeSpan.Zero);
>
> Is there any way or practice (besides asking i.e. on StackOverflow) I can
> perform so I can know that?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/OpenRA/OpenRA/pull/7358/files#r23191774>.
>